compiler/compile/OMRSymbolReferenceTable.hpp

```diff
@@ -206,6 +206,16 @@ class SymbolReferenceTable
       startPCLinkageInfoSymbol,
       instanceShapeFromROMClassSymbol,
 
+      /** \brief Performs a substitutability comparison between two objects.
+       *
+       * The comparison takes as arguments references to the two objects.
+       *
+       * If the two objects are reference types, this is equivalen to a pointer compare.
+       * If the two objects are value types, then a structural comparison betweeen the two is performed.
+       * If the two objects are of different types, the exact behaviour depends on the language.
```

Comment:
Are the above semantics strictly implied for OMR or is this simply a helper to represent a comparison that may or may not be a trivial pointer compare. The language would then supply the semantics by implementing the helper and the appropriate evaluators etc?

Reply:
That's a good question @andrewcraik.

My intention was to define semantics for the `substitutabilityComparison` non-helper that are general enough to be useful in multiple languages while being specific enough to enable useful optimizations.

I'm not sure if I achieved that intention successfully so I'm definitely open to suggestions for improvement.

My reasoning was that decaying the operation from a non-helper call to a simple acmp could be a generic optimization in OMR. However, I'm not sure whether it would be useful. If not, then it might make sense to relax the semantics and give language implementations the freedom to chose how non-helper calls are handled.

Reply:
The decay seems quite language specific to me, but I would value an opinion from @mstoodle on where he feels the line might be - having a decent implementation in OMR makes sense, but this seems like a language semantics extension point to me (at least conceptually).

Reply:
Feels to me like this symbol provides a language with a way to compute a richer form of object equality and we should probably leave it at that.  "Value types" as such don't really exist in OMR, unless I missed something, so outlining semantics in terms of objects and value types may only serve to confuse people. The exact conditions under which a particular language runtime can use reference equality (i.e. `acmpeq`/`acmpne`) for object equality should be left to the language runtime to decide. The mechanics to transform the call with the symbol into the address compare node are reasonably simple. All that as a longwinded explanation for: I think I'm with @andrewcraik on this one.

Reply:
Alright, fair enough. We should pick a different name for the symbol then; one that better conveys the "richer form of object equality." Any suggestions? `richEqualityComparison` sounds a little off to me.

Reply:
how about `objectEqualityComparison`?

Reply:
Ok, I think I can work with that.

Reply:
or even just `compareObjectEquality` :)

Reply:
Fixed.